File: course/section_4_stack/code/11-remove-k-digits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        if len(num) <= k:
            return "0"
        stack = []
        for c in num:
            while k > 0 and stack and stack[-1] > c:
                stack.pop()
                k -= 1
            stack.append(c)
        stack = stack[: len(stack) - k]
        while stack and stack[0] == "0":
            stack.pop(0)
        res = "".join(stack)
        return res if res else "0"

File: course/section_4_stack/code/11-remove-k-digits.php
<?php

class Solution {
    function removeKdigits($num, $k) {
        $stack = [] ; 
        for($i = 0 ; $i < strlen($num); $i++) {
            $c = $num[$i] ; 
            while($k > 0 && !empty($stack) && end($stack) > $c){
                array_pop($stack);
                $k-- ;
            }
            $stack[] = $c ;
        }
        while($k > 0){
            array_pop($stack);
            $k--;
        }
        $res = ltrim(implode('', $stack) , '0');
        return $res === '' ? '0' : $res ;
    }
}